### PR TITLE
Fix reuse of invalid write cursor

### LINKED ISSUE #9913: tikv keep panicking at components/engine_rocks/src/engine_iterator.rs:39
### What version of TiKV are you using?
Master.

### What operating system and CPU are you using?
Doesn't matter.

### Steps to reproduce
Not sure.

### What did you expect?
TiKV keeps running without panicking.

### What did happened?
TiKV panics:
```
[2021/03/29 12:13:21.004 +08:00] [FATAL] [lib.rs:455] ["assertion failed: cfg!(feature = \\\"nortcheck\\\") || self.valid()?"] [backtrace="stack backtrace:\n   0: t[2/9902]::set_panic_hook::{{closure}}\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_util/src/lib.rs:454\n   1: std::panicking::rust_panic_with_hook\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:595\n   2: std::panicking::begin_panic_handler::{{closure}}\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:495\n   3: std::sys_common::backtrace::__rust_end_short_backtrace\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/sys_common/backtrace.rs:141\n   4: rust_begin_unwind\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/std/src/panicking.rs:493\n   5: core::panicking::panic_fmt\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/panicking.rs:92\n   6: core::panicking::panic\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35//library/core/src/panicking.rs:50\n   7: <engine_rocks::engine_iterator::RocksEngineIterator as engine_traits::iterable::Iterator>::next\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/engine_rocks/src/engine_iterator.rs:39\n   8: raftstore::store::region_snapshot::RegionIterator<S>::next\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/raftstore/src/store/region_snapshot.rs:339\n      tikv_kv::raftstore_impls::<impl tikv_kv::Iterator for raftstore::store::region_snapshot::RegionIterator<S>>::next\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_kv/src/raftstore_impls.rs:86\n      tikv_kv::cursor::Cursor<I>::next\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_kv/src/cursor.rs:367\n      tikv::storage::txn::actions::get_old_value::get_old_value\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/actions/get_old_value.rs:34\n      tikv::storage::txn::actions::prewrite::prewrite\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/actions/prewrite.rs:82\n   9: tikv::storage::txn::commands::prewrite::Prewriter<K>::prewrite\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/commands/prewrite.rs:445\n      tikv::storage::txn::commands::prewrite::Prewriter<K>::process_write\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/commands/prewrite.rs:370\n      <tikv::storage::txn::commands::prewrite::Prewrite as tikv::storage::txn::commands::WriteCommand<S,L>>::process_write\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/commands/prewrite.rs:198\n  10: tikv::storage::txn::commands::Command::process_write\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/commands/mod.rs:539\n  11: tikv::storage::txn::scheduler::Scheduler<E,L>::process_write\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/scheduler.rs:635\n      tikv::storage::txn::scheduler::Scheduler<E,L>::process_by_worker::{{closure}}::{{closure}}\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/scheduler.rs:579\n      tikv_kv::with_tls_engine::{{closure}}\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_kv/src/lib.rs:401\n      std::thread::local::LocalKey<T>::try_with\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/thread/local.rs:272\n      std::thread::local::LocalKey<T>::with\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/thread/local.rs:248\n      tikv_kv::with_tls_engine\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_kv/src/lib.rs:399\n      tikv::storage::txn::scheduler::Scheduler<E,L>::process_by_worker::{{closure}}\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/src/storage/txn/scheduler.rs:578\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/future/mod.rs:80\n      tikv_util::yatp_pool::future_pool::FuturePool::spawn::{{closure}}\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_util/src/yatp_pool/future_pool.rs:113\n      <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/core/src/future/mod.rs:80\n  12: <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/task/future.rs:261\n  13: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle\n             at /home/jenkins/agent/workspace/tikv_ghpr_build_release/tikv/components/tikv_util/src/yatp_pool/mod.rs:93\n      yatp::pool::worker::WorkerThread<T,R>::run\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/pool/worker.rs:48\n      yatp::pool::builder::LazyBuilder<T>::build::{{closure}}\n             at /rust/git/checkouts/yatp-e704b73c3ee279b6/6bbea16/src/pool/builder.rs:91\n      std::sys_common::backtrace::__rust_begin_short_backtrace\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/sys_common/backtrace.rs:125\n  14: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/thread/mod.rs:474\n      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/panic.rs:322\n      std::panicking::try::do_call\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/panicking.rs:379\n      std::panicking::try\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/panicking.rs:343\n      std::panic::catch_unwind\n             at /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/panic.rs:396\n      std::thread::Builder::spawn_unchecked::{{closure}}\n             at
 /rustc/bc39d4d9c514e5fdb40a5782e6ca08924f979c35/library/std/src/thread/mod.rs:473\n      core::ops::function::FnOnce::call_once{{vtable.shim[2021/03/29 12:13:36.623 +08:00
```

Seems it's introduced in #9630 . However maybe we need to post a branch to get the error returned by `DB::valid` to locate the reason.

### PR DESCRIPTION
**Pull request by tabokie**
### What problem does this PR solve?

Issue Number: close #9913

write_cursor could be invalidated during `check_data_constraint` routine, and reused by `get_old_value` which assumes the cursor is unchanged.

### What is changed and how it works?

What's Changed:

Remove the usage of write_cursor in function `get_old_value`.

~I'm proposing two potential fixes for this issue:
a) Creating new iterator for `check_data_constraint`
b) Add additional check to safely reuse the could-be invalid cursor, replying on the fact that `check_data_constraint` seeks in the same direction and terminates at the same condition as `get_old_value`.
Those two fixes are both included here as two separate commits.~

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

**Review by Connor1996 — COMMENTED** [review]
Please add a test

**Comment by tabokie** [comment]
After discussed offline with @5kbpers , we decided to remove this specific branch in `get_old_value`. Fix 1 adds unnecessary overhead for every transaction. Fix 2 creates a painful coupling between the two functions' logic.

File: src/storage/txn/actions/get_old_value.rs [hunk 1]
```diff
@@ -9,39 +9,14 @@ use txn_types::{Key, OldValue, Write};
...
-        let write_under_cursor =
-            txn_types::WriteRef::parse(cursor.value(&mut txn.reader.statistics.write))
-                .unwrap()
-                .to_owned();
-        key.clone() == key_under_cursor && w.clone() == write_under_cursor
-    } else {
-        true
-    });
     match prev_write {
-        Some(w) if !w.may_have_old_value() => {
```
Comment (Connor1996) [inline src/storage/txn/actions/get_old_value.rs:31]: So how do we handle the case now? it doesn't affect correctness?

**Review by Connor1996 — COMMENTED** [review]
Please add a test

File: src/storage/txn/actions/get_old_value.rs (hunk 1 above)
Comment (5kbpers) [reply src/storage/txn/actions/get_old_value.rs:31]: cdc worker will try to get old value from rocksdb when `prev_write` is `Rollback` or `Lock` type

Comment (5kbpers) [reply src/storage/txn/actions/get_old_value.rs:31]: No much effect because those cases should be rare.

**Review by youjiali1995 — COMMENTED** [review]
Code changes lgtm

**Review by Connor1996 — APPROVED** [review]
LGTM

File: src/storage/txn/actions/get_old_value.rs [hunk 2]
```diff
@@ -9,39 +9,14 @@ use txn_types::{Key, OldValue, Write};
 /// it must be read in the caller and be passed in for optimization
 pub fn get_old_value<S: Snapshot>(
     txn: &mut MvccTxn<S>,
-    key: &Key,
+    _key: &Key,
```
Comment (Connor1996) [inline src/storage/txn/actions/get_old_value.rs:12]: remove it？

**Review by 5kbpers — APPROVED** [review]

**Pull request merged**